Sources/Moya/TargetType.swift

```diff
@@ -38,32 +32,39 @@ public extension TargetType {
     }
 }
 
-/// Represents a type of upload task.
-public enum UploadType {
+/// Represents an HTTP task.
+public enum Task {
 
-    /// Upload a file.
-    case file(URL)
+    /// A request with no additional data.
+    case requestPlain
 
-    /// Upload "multipart/form-data"
-    case multipart([MultipartFormData])
-}
+    /// A requests body set with data.
+    case requestData(Data)
 
-/// Represents a type of download task.
-public enum DownloadType {
+    /// A requests body set with encoded parameters.
+    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
 
-    /// Download a file to a destination.
-    case request(DownloadDestination)
-}
+    /// A requests body set with data, combined with url parameters.
+    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
 
-/// Represents an HTTP task.
-public enum Task {
+    /// A requests body set with encoded parameters combined with url parameters.
+    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
+
+    /// A file upload task.
+    case uploadFile(URL)
+
+    /// A file upload task with extra parameters using the given encoding.
+    case uploadFileParameters(parameters: [String: Any], encoding: ParameterEncoding, URL)
+
+    /// A "multipart/form-data" upload task.
+    case uploadMultipart([MultipartFormData])
 
-    /// A basic request task.
-    case request
+    /// A "multipart/form-data" upload task with extra parameters using the given encoding.
+    case uploadMultipartParameters(parameters: [String: Any], encoding: ParameterEncoding, [MultipartFormData])
```

Comment:
Correct me if I'm wrong, but if you want to add parameters to the multipart request, you want to add them using `MultipartFormData`, not `parameters` property (at least in Alamofire). You may want to add parameters to the URL, though, so that would be an exception. But I'm not sure if we should allow more encodings, like JSON, too.